File: MidtermLab1/sales_summary_dashboard.js
const fs = require('fs');
const readline = require('readline');
const path = require('path');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askFilePath() {
    rl.question("Enter dataset file path: ", function(filePath) {
        if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
            console.log("File found. Processing...\n");
            const analyzer = new SalesAnalyzer(filePath);
            analyzer.processData()
                .then(() => {
                    analyzer.displayReport();
                    rl.close();
                })
                .catch(error => {
                    console.error("Error processing file: " + error.message);
                    rl.close();
                });
        } else {
            console.log("Invalid file path. Try again.");
            askFilePath();
        }
    });
}

class SalesAnalyzer {
    constructor(filePath) {
        this.filePath = filePath;
        this.records = [];
        this.totalSales = 0;
        this.highestSale = Number.NEGATIVE_INFINITY;
        this.lowestSale = Number.POSITIVE_INFINITY;
    }

    async processData() {
        return new Promise((resolve, reject) => {
            const stream = fs.createReadStream(this.filePath, { encoding: 'utf8' });
            let buffer = '';
            let isFirstLine = true;

            stream.on('data', chunk => {
                buffer += chunk;
                let idx;
                while ((idx = buffer.indexOf('\n')) >= 0) {
                    const line = buffer.slice(0, idx).replace(/\r$/, '');
                    buffer = buffer.slice(idx + 1);

                    if (isFirstLine) {
                        isFirstLine = false; // skip header
                    } else {
                        this.processLine(line);
                    }
                }
            });

            stream.on('end', () => {
                if (buffer.trim()) {
                    if (!isFirstLine) this.processLine(buffer.replace(/\r$/, ''));
                }
                resolve();
            });

            stream.on('error', reject);
        });
    }

    processLine(line) {
        if (!line || !line.trim()) return;

        const parts = parseCSVLine(line);
        if (parts.length > 7) {
            const totalSalesStr = (parts[7] || '').trim();
            if (totalSalesStr) {
                const sales = parseFloat(totalSalesStr);
                if (!isNaN(sales)) {
                    this.records.push({
                        title: parts[1] || '',
                        console: parts[2] || '',
                        sales: sales
                    });
                    this.totalSales += sales;
                    if (sales > this.highestSale) this.highestSale = sales;
                    if (sales < this.lowestSale) this.lowestSale = sales;
                }
            }
        }
    }

    displayReport() {
        const fileName = path.basename(this.filePath);
        const totalRecords = this.records.length;
        const averageSales = totalRecords > 0 ? this.totalSales / totalRecords : 0;

        console.log('='.repeat(60));
        console.log('               SALES SUMMARY DASHBOARD');
        console.log('='.repeat(60));
        console.log(`\nDataset File: ${fileName}\n`);
        console.log('--- ANALYTICS REPORT ---\n');
        console.log(`Total Number of Records: ${totalRecords}`);
        console.log(`Total Sales/Revenue:     $${this.totalSales.toFixed(2)} million`);
        console.log(`Average Sales per TX:    $${averageSales.toFixed(2)} million`);

        if (totalRecords > 0) {
            console.log(`Highest Single TX:       $${this.highestSale.toFixed(2)} million`);
            console.log(`Lowest Single TX:        $${this.lowestSale.toFixed(2)} million`);
        } else {
            console.log('Highest Single TX:       N/A');
            console.log('Lowest Single TX:        N/A');
        }

        console.log('\n' + '='.repeat(60));
    }
}

// Basic CSV parser handling quotes and escaped quotes
function parseCSVLine(line) {
    const result = [];
    if (line == null) return result;
    let cur = '';
    let inQuotes = false;
    for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (ch === '"') {
            if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
                cur += '"'; i++; // escaped quote
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === ',' && !inQuotes) {
            result.push(cur);
            cur = '';
        } else {
            cur += ch;
        }
    }
    result.push(cur);
    return result;
}

// Start
askFilePath();

File: MidtermLab1/SalesSummaryDashboard.java
import java.io.*;
import java.util.*;

public class SalesSummaryDashboard {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        File file;

        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = input.nextLine();
            file = new File(path);
            if (file.exists() && file.isFile()) {
                break;
            } else {
                System.out.println("Invalid file path. Please try again.");
            }
        }

        try {
            SalesAnalyzer analyzer = new SalesAnalyzer(file);
            analyzer.processData();
            analyzer.displayReport();
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        } finally {
            input.close();
        }
    }
}

class SalesAnalyzer {
    private File file;
    private List<SalesRecord> records;
    private int totalRecords;
    private double totalSales;
    private double averageSales;
    private double highestSale;
    private double lowestSale;

    public SalesAnalyzer(File file) {
        this.file = file;
        this.records = new ArrayList<>();
        this.totalSales = 0;
        this.highestSale = Double.NEGATIVE_INFINITY;
        this.lowestSale = Double.POSITIVE_INFINITY;
    }

    public void processData() throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            String line;
            reader.readLine(); // skip header

            while ((line = reader.readLine()) != null) {
                if (line.trim().isEmpty()) continue;

                String[] parts = line.split(",");
                if (parts.length > 2) {
                    try {
                        double sales = Double.parseDouble(parts[2].trim());

                        String title = parts[0].trim();
                        String console = parts[1].trim();

                        records.add(new SalesRecord(title, console, sales));
                        totalSales += sales;

                        if (sales > highestSale) highestSale = sales;
                        if (sales < lowestSale) lowestSale = sales;
                    } catch (NumberFormatException ignored) {
                        // skip invalid data safely
                    }
                }
            }
        }

        totalRecords = records.size();
        averageSales = totalRecords > 0 ? totalSales / totalRecords : 0;
    }

    public void displayReport() {
        System.out.println("\n============================================================");
        System.out.println("                     SALES SUMMARY DASHBOARD");
        System.out.println("============================================================");
        System.out.println("File: " + file.getName());
        System.out.println("------------------------------------------------------------");
        System.out.printf("Total Records:            %d%n", totalRecords);
        System.out.printf("Total Sales/Revenue:     $%.2f%n", totalSales);
        System.out.printf("Average Sales/Transaction:$%.2f%n", averageSales);

        if (totalRecords > 0) {
            System.out.printf("Highest Sale:            $%.2f%n", highestSale);
            System.out.printf("Lowest Sale:             $%.2f%n", lowestSale);
        } else {
            System.out.println("Highest Sale:            N/A");
            System.out.println("Lowest Sale:             N/A");
        }
        System.out.println("============================================================");
    }
}

class SalesRecord {
    private String title;
    private String console;
    public SalesRecord(String title, String console, double sales) {
        this.title = title;
        this.console = console;
    }

    public String getTitle() {
        return title;
    }

    public String getConsole() {
        return console;
    }
}